solutions/cf-1702E.mdx

```diff
@@ -0,0 +1,129 @@
+$a_i$ and $b_i$. Then we draw an edge between dominoes $i$ and $j$ if and only if $a_i = a_j$ or $a_i = b_j$ or $a_j = b_i$
+or $a_j = b_j$.
+
+Now, suppose that we have a way to create the 2 sets, with $\frac{n}2$ dominoes in each set.
+Since each number appears only once in each set, this graph is bipartite.
+
+This leads us to the following assertion: It is possible to construct the two sets
+[if and only if](https://artofproblemsolving.com/wiki/index.php/Iff) the graph constructed as above is bipartite. We proved
+the only if clause above, all that remains is to prove the if clause.
+
+Suppose the graph is bipartite. Divide the graph into two sets according to the bipartite split of the graph. We wish to prove
+that no numbers are repeated twice in either of these sets. However, since an edge exists between any two dominoes which share
+a number, it is impossible for two dominoes which share a number to be in the same set (if it were possible, we would have a
+cycle of length 1, which is odd, contradicting the fact that the graph is bipartite). So, given a bipartite split of the graph,
+we can easliy construct our 2 desired sets.
+
+The implementation of this is relatively straight forward; Simply check if the graph is bipartite or not, and output accordingly.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+bool dfs(int u, int g, vector<int>& vis, vector<vector<int>>& adj) {
+	// If the node is already visited with and was marked in the same color
+	if (vis[u] == g) {
+		return true;
+	}
+    
+	// If the node is already visited with and was marked in the other color
+	if (vis[u] == 1 - g) {
+		return false;
+	}
+    
+	vis[u] = g; 
+	for (auto i : adj[u]) {
+        	// If there is a problem when dfsing to any of the other nodes we 
+        	// need to return false
+		if (!dfs(i, 1 - g, vis, adj)) return false; 
+	}
+	// No problems
+	return true;
+}
+
+void solve() {
+	int n; 
+	cin >> n;
+	vector<pair<int, int>> dm(n);
+    
+	for (auto& i : dm) {
+		cin >> i.first >> i.second;
+	}
+		
+	for (auto i : dm) {
+        	// Check if a dominoe has the same number twice
+		if (i.first == i.second) {
+			cout << "NO\n";
+			return;
+		}
+	}
+
+	// Check if a number exists in 3 or more dominoes
+	map<int, vector<int>> ma;
+	for (int i = 0; i < n; i++) {
+		ma[dm[i].first].push_back(i);
+		ma[dm[i].second].push_back(i);
+		if (ma[dm[i].first].size() > 2 || ma[dm[i].second].size() > 2) {
+			cout << "NO\n";
+			return;
+		}
+	}
+
+	// initialize variables for DFS
+	vector<int> vis(n, -1);
+	vector<vector<int>> adj(n);
+	for (auto p : ma) {
+		auto v = p.second;
+		if (v.size()==2) {
+			adj[v[0]].push_back(v[1]);
+			adj[v[1]].push_back(v[0]);
+		}
+	}
+	
+	/**
+	 * Check if the graph is bipartite:
+	 * Since the graph is not guaranteed to be connected
+	 * we need to iterate over every value and make sure
+	 * every value is visited
+	**/
```

Comment:
```suggestion
	/*
	 * Check if the graph is bipartite:
	 * Since the graph is not guaranteed to be connected
	 * we need to iterate over every value and make sure
	 * every value is visited
	*/
```